--- numpy/distutils/system_info.py.orig	2011-08-08 11:40:37.514881710 -0500
+++ numpy/distutils/system_info.py	2011-08-08 11:41:36.386872112 -0500
@@ -1015,9 +1015,6 @@
             else:
                 info['language'] = 'f77'
 
-        atlas_version, atlas_extra_info = get_atlas_version(**atlas)
-        dict_append(info, **atlas_extra_info)
-
         self.set_info(**info)
 
 class atlas_blas_info(atlas_info):
@@ -1042,9 +1039,6 @@
             dict_append(info,include_dirs=[h])
         info['language'] = 'c'
 
-        atlas_version, atlas_extra_info = get_atlas_version(**atlas)
-        dict_append(atlas, **atlas_extra_info)
-
         dict_append(info,**atlas)
 
         self.set_info(**info)
