--- python-updater.orgi	2010-08-05 09:53:01.000000000 -0500
+++ python-updater	2010-08-05 09:58:20.000000000 -0500
@@ -24,7 +24,7 @@
 #
 
 VERSION="0.7"
-NEW_PY_VER=$(/usr/bin/python -V 2>&1 | sed 's:Python ::' | cut -d. -f1-2)
+NEW_PY_VER=$("@GENTOO_PORTAGE_EPREFIX@"/usr/bin/python -V 2>&1 | sed 's:Python ::' | cut -d. -f1-2)
 
 PKGS_EXCEPTIONS="dev-lang/python sys-apps/portage"
 PKGS_MANUAL="app-office/gnumeric app-office/dia dev-libs/boost x11-libs/vte"
@@ -53,10 +53,10 @@
 # load the gentoo-style info macros, but hack to get around
 # it thinking this is an rc script
 EBUILD="1"
-source /etc/init.d/functions.sh
+source "@GENTOO_PORTAGE_EPREFIX@"/etc/init.d/functions.sh
 
 # portage variables
-PKG_DBDIR=/var/db/pkg
+PKG_DBDIR="@GENTOO_PORTAGE_EPREFIX@"/var/db/pkg
 
 # usage()
 # display usage
@@ -120,7 +120,7 @@
 # Find old python version, return non-zero if not found
 get_old_pyver() {
     for old in ${PYTHON_VERSIONS}; do
-        if [[ "${old}" != "${NEW_PY_VER}" && -e /usr/bin/python${old} ]]
+        if [[ "${old}" != "${NEW_PY_VER}" && -e "@GENTOO_PORTAGE_EPREFIX@"/usr/bin/python${old} ]]
         then
                 echo -n "${old}"
                 return 0
@@ -134,7 +134,7 @@
 # get_portage_portdir()
 # Check if portage knows about PORTDIR and return it
 get_portage_portdir() {
-    local portdir="$(/usr/bin/portageq portdir)"
+    local portdir="$("@GENTOO_PORTAGE_EPREFIX@"/usr/bin/portageq portdir)"
 
     if [[ -z "${portdir}" ]]; then
         eerror "Unable to proceed. Can not find PORTDIR. Make sure the command:"
@@ -281,7 +281,9 @@
         CHECK_SONAME=0
     else
         veinfo 1 'check "soname" enabled.'
-        OLD_SONAME="$(readlink -n /usr/lib/libpython${OLD_PY_VER}.so)"
+        OLD_SONAME="@GENTOO_PORTAGE_EPREFIX@/usr/lib/libpython${OLD_PY_VER}@GENTOO_PORTAGE_LIBSUFFIX@"
+        [[ -e "${OLD_SONAME}" ]] || OLD_SONAME=""
+        [[ -L "${OLD_SONAME}" ]] && OLD_SONAME="$(readlink -n ${OLD_SONAME})"
         if [[ -z "${OLD_SONAME}" ]]; then
             ewarn "Couldn't find old libpython soname"
             ewarn "Disabling soname check."
@@ -372,7 +374,7 @@
         # /usr/lib/python$old
         # /usr/lib32/python$old
         # /usr/lib64/python$old
-        if grep -qe "/usr/\(include\|lib\(32\|64\)\?\)/python${OLD_PY_VER}" ${content}; then
+        if grep -qe "@GENTOO_PORTAGE_EPREFIX@/usr/\(include\|lib\(32\|64\)\?\)/python${OLD_PY_VER}" ${content}; then
             PKGS_TO_REMERGE="${PKGS_TO_REMERGE} ${CATPKGVER}"
             eindent
             einfo "Adding to list: ${CATPKGVER}"
